### DISCUSSION TITLE
named parameters

### DISCUSSION THREAD

**Discussion by ghost** [powersync-ja/sqlite_async.dart#56] [Ideas]
would it be possible to expose executing queries with named parameters? sqlite3 seems to [have the feature](https://github.com/simolus3/sqlite3.dart/issues/63)

**Comment by rkistner** [comment]
It should be possible - we'll just have to figure out a good API for exposing that.

**Reply by ryanheise** [reply]
Would this API tick the boxes?

```dart
resultSet = await getAllWith(
  'SELECT * from mytable WHERE id = :id',
  StatementParameters.named({':id': 3}),
);
```

This is inspired by the sqlite3 package's API by adding the `With` suffix whenever a `StatementParameters` parameter is accepted. The inclusion of the colon in the parameter key is not to my liking, but that's reflective of the way the C API works, and also the way the sqlite3 package works.

Not sure if you would want to just reuse sqlite3's `StatementParameters` class directly since it also includes a constructor for custom bindings which exposes the `PreparedStatement` API outside of the isolate. But you could at least invent your own equivalent of `StatementParameters` unifying named and indexed parameters.

Another approach could be:

```dart
resultSet = await getAllNamed(
  'SELECT * from mytable WHERE id = :id',
  parameters: {'id': 3},
);
```

In hindsight, if one does not want a separate version of each method to support different parameter binding approaches, then maybe repurposing the original methods to look like the first approach:

```dart
// for named
resultSet = await getAll(
  'SELECT * from mytable WHERE id = :id',
  Parameters.named({':id': 3}),
);
// for indexed
resultSet = await getAll(
  'SELECT * from mytable WHERE id = ?',
  Parameters([3]),
);
```

If Dart had union types, then you could also just keep the API the same as it is now and accept either a List or a Map as a parameter (I suppose that is possible with a `dynamic` or `Object?` parameter and runtime type checking.)
